### DISCUSSION TITLE
Help us to choose minimum required C++ standard for OpenCV 5.0

### DISCUSSION THREAD

**Discussion by vpisarev** [opencv/opencv#25039] [Polls]
In OpenCV 4.x the minimum required standard is C++ 11. For OpenCV 5.0 we are going to raise the bar.  Note that in the subsequent 5.x releases we could raise it up again, but not down. For your reference, please, check C++ standard compliance status in different mainstream compilers (scroll to the bottom to see C++14 and C++17 support status): https://en.cppreference.com/w/cpp/compiler_support

**Comment by fengyuentau** [comment]
Mainstream compilers (gcc, clang, msvc) have almost complete support of C++17 (Some are missing one or two item though). Also `std::string_view` is available since C++17 since it is needed in a evolution proposal.

Vote for C++17 🙋

**Comment by ShiqiYu** [comment]
My only concern about C++17 is some low-cost embedded systems. GCC7 is needed for C++17, and GCC7 was released in 2017. GCC 7.4 is in Ubuntu 18.04. I think very few systems still use an earlier version of Ubuntu 18.04. I think C++17 is possible for OpenCV 5.

**Comment by edvn0** [comment]
Vote for C++17!

**Comment by fspindle** [comment]
I'm involved in a number of software developments that are intended to be embedded in computers on board satellites and for which c++14 or later standards are not authorised. I therefore vote in favour of maintaining a minimum c++11 standard.

**Reply by LaurentBerger** [reply]
Do you ask for a c++11 checkbox?

**Reply by mshabunin** [reply]
I believe it is possible to define compilation standard by default, but avoid using parts of language and standard library which were not supported or fully implemented. For example, libstdc++ states full support of `memory_resource` since version 9.1 (https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017), but older compiler, like GCC 7.5.0 (Ubuntu 18.04) already supported `--std=c++17` option. We can avoid using `memory_resource` and remain compatible with older compilers with partial C++17 support.

Compiler versions in some Linux OSes:
* Ubuntu 18.04 - GCC 7.5, Clang 6.0 (EOL Apr 2023)
* Ubuntu 20.04 - GCC 9.4, Clang 10
* Ubuntu 22.04 - GCC 11.4 Clang 14
* Debian 10 - GCC 8.3, Clang 7
* Rocky Linux 8 (CentOS?, RHEL) - GCC 8.5, Clang 16
* NVIDIA Jetson Linux - GCC 9.3 (?)

So we can target C++17, but also check build on Ubuntu 18.04 to make sure older toolchains with incomplete standard support can build OpenCV. Additionally, modern modules can require more features or higher C++ standard and turn themselves off automatically if compiler does not satisfy their requirements. Example: https://github.com/opencv/opencv_contrib/blob/4.x/modules/cvv/CMakeLists.txt#L1

It would be nice to see examples of older toolchains/platforms which are currently used by people to buld OpenCV to take them in account too.

**Reply by bebuch** [reply]
@fspindle Is there a reason that an old compiler is combined with the latest version of OpenCV on these devices? Would it be an option to simply continue using OpenCV 4 there? To be honest, I am very skeptical that the combination of an extremely old compiler and top current libraries will really be used productively anywhere.

@mshabunin I think it would be better to set complete C++17 as the lower limit. If certain features are not allowed to be used, then it becomes very difficult for the developers to decide what exactly they are allowed to use and what not.

**Reply by fspindle** [reply]
@bebuch Certainly, continuing to use OpenCV 4 when older compilers are needed is an interesting option that could be considered. I agree that enabling/disabling features, whether available or not, is a pain for developers. Targetting C++17 becomes a good compromise.

**Reply by mshabunin** [reply]
> I think it would be better to set complete C++17 as the lower limit.

In this case we would need to explicitly forbid compilers/libraries older than some specific version. According to [GCC](https://gcc.gnu.org/projects/cxx-status.html)/[libstdc++](https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017) documentation it would be GCC 12. And Microsoft [documentation](https://learn.microsoft.com/en-gb/cpp/overview/visual-cpp-language-conformance) states Visual Studio 2019 16.4. It would mean that we won't support even Ubuntu 22 (GCC 11). I believe these are too strong requirements.

I thought we would set baseline compiler versions to something like GCC 9 and VS 2017 15.3 and would check these builds on CI. Perhaps also emit a warning in cmake when building with older versions of mainstream compilers.

> it becomes very difficult for the developers to decide what exactly they are allowed to use and what not

I wouldn't say so. Algorithm for using some feature would be as follows:
- is this feature already used in OpenCV?
- is this feature part of C++17 standard?
- is this feature supported by minimal compiler versions set by OpenCV? (can be checked in the Compiler Explorer or locally in container/VM)
- does your (preliminary) patch pass OpenCV CI?

Even then, some more advanced features can still be used in optional cases (e.g. in samples) under conditional compilation.

**Comment by sabudilovskiy** [comment]
Where c++20?

**Reply by vpisarev** [reply]
@sabudilovskiy, sorry, we only offer to choose from the feasible for us options :) C++ 20 introduces *a lot* of great stuff, and some of it we could definitely use, like "modules", but not all of the features, including modules, are properly implemented in all mainstream compilers. Many of the features are implemented, but require quite fresh versions of compilers that users may not have installed by default, see the detailed post above.

For example, GCC, the default C++ compiler for Linux distributions, currently declares full support of C++ 17 (whereas C++ 20 support is still partial and is described as "experimental") and the default standard is "gnu++17" (C++ 17 with GCC extensions) since GCC 11: https://gcc.gnu.org/projects/cxx-status.html. C++ 14 (well, gnu++14) has been the default standard from GCC 6 till GCC 10.x inclusive (but, as Maksim pointed out, starting from GCC 8.x it's quite safe to use C++ 17 except for a few missing libstdc++ features).

Let's wait for a few years while compilers and distro's will catch up. Maybe then we will make another leapfrog straight to C++ 23 (or even C++ 26).

**Comment by LaurentBerger** [comment]
@vpisarev What's opencv team choice?

**Comment by Ginkgo-Biloba** [comment]
Vote C++17.
Some old gcc has trouble with C++14 `constexpr` function. When I compiler OpenCV / spdlog with gcc 5.x, it will issue errors like `constexpr function call non-constexpr function` in protobuf / fmtlib, so I have to delete `constexpr` / use C++11. Also https://stackoverflow.com/questions/54326901/why-gcc5-4-doesnt-compile-a-constexpr-function-calling-a-non-constexpr-function

**Comment by aliosmanuysal** [comment]
❤️

**Comment by aferust** [comment]
> In OpenCV 4.x the minimum required standard is C++ 11. For OpenCV 5.0 we are going to raise the bar. Note that in the subsequent 5.x releases we could raise it up again, but not down. For your reference, please, check C++ standard compliance status in different mainstream compilers (scroll to the bottom to see C++14 and C++17 support status): https://en.cppreference.com/w/cpp/compiler_support

C++20 with modules. Yeah this is a daydream. And please also contribute to DCV., written in a sane language dlang. https://github.com/libmir/dcv

**Comment by bebuch** [comment]
IMO, developers who are stuck with a very old compiler version can continue to use OpenCV 4. New library versions should target new compiler versions. C++17 is already a conservative choice.

**Comment by dkurt** [comment]
Regarding C++17, we can use [std::any](https://en.cppreference.com/w/cpp/utility/any) for any-value map as a replacement for [cv::dnn::Dict](https://docs.opencv.org/5.x/d9/d2b/classcv_1_1dnn_1_1Dict.html) or [FileNode](https://docs.opencv.org/5.x/de/dd9/classcv_1_1FileNode.html)

**Comment by gblikas** [comment]
If I had to choose between the two, I'd pick C++17. 

_However_... It may be an unpopular opinion but I think OpenCV's documentation isn't ready for an upgrade past C++11. At the moment, there are so many modules with varying levels of documentation. I believe that without an improvement to the contribution guidelines, upgrading the C++ Standard will cause what @mshabunin is talking about. There is very little documentation of developers to figure out why their compilations may be failing (like in older embedded systems, as @fspindle points out), or what functions developers agree to (not) use.

**Comment by shengyu7697** [comment]
Vote for C++17

**Comment by JWCS** [comment]
A small comment on future decisions: every 3-7 years enough time has passed for the new relevant version of C++ to have matured with various compilers. Furthermore, each version is with increasingly with interesting or critical features, but not very backwards compatible. We should just pre-emptively decide to release a new major version of opencv with each C++ revision, fully embracing the changes within. So OpenCV 5 can be C++17 based, and should fully leverage whichever nice new api features it can. And when C++20 is deemed stable enough with compilers, we should release an OpenCV 6, that fully embraces modules (which changes everything about compilation times) (and this would be a major change).

**Comment by YunKaiRiYueLang** [comment]
May be c++11?

c++17 is better than c++14;  I can't vote c++14; But actually,  There are so many  device on board do not support c++17.

**Reply by bebuch** [reply]
Is there a reason that an outdated compiler is combined with the latest version of OpenCV on these devices? Would it be an option to simply continue using OpenCV 4 there?

**Reply by YunKaiRiYueLang** [reply]
Some ARM architecture servers used for AI program .    The dnn module  used very frequently.I use dnn module in opencv4 now. With other Edge devices with low computing power,I can use opencv3 or opencv4.This is why.

**Comment by Dharmesh946** [comment]
By default I voted for C++17 but why not C++20 that is now very well supported?

**Reply by bebuch** [reply]
The support for C++20 is still not complete for all major compilers and standard libraries. Additionally we should support the latest 2 or 3 versions of every compiler so it works by default for the most Linux distributions that are still in support cycle.

**Comment by mshabunin** [comment]
Please check preliminary PR establishing C++17 baseline: #26287. We will be glad to hear your comments and concerns.
